### PR TITLE
Fix NPE when using error binding pattern with no message

### LINKED ISSUE #28020: Error binding pattern without error message gives NullPointerException
**Description:**
case 1:
```bal
type myDetails record {
    int a;
};
type myError error<myDetails>;
public function main() {
    myError error(a = mya) = error myError("message", a = 5);
}
```
gives 
```
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.validateErrorVariable(SymbolEnter.java:2410)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.symbolEnterAndValidateErrorVariable(SymbolEnter.java:2370)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1026)
	at org.wso2.ballerinalang.compiler.tree.BLangErrorVariable.accept(BLangErrorVariable.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2756)
```
case 2: 
```bal
type Error error<record {int code; boolean fatal;}>;

public function testBasic() {
    int i;
    boolean b;
    error(code = i, fatal = b) = error Error("message", code = 100, fatal = true);
}
```
gives error
`missing error message binding pattern`

**Spec allows error binding patterns without the message**

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title.

Fixes #28020

## Approach
According to the spec error binding pattern without the error message should be allowed.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -1581,6 +1581,10 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable) {
         errorVariable.type = errorType;
 
         if (!errorVariable.isInMatchStmt) {
+            if (errorVariable.message == null) {
+                dlog.error(errorVariable.pos, DiagnosticErrorCode.INVALID_ERROR_BINDING_PATTERN, errorVariable.type);
+                return false;
+            }
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1587]: Spec allows error binding patterns without the message

**Review by rdulmina — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 1 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Fixed in 8aaa9bc

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -1995,14 +1999,14 @@ public void visit(BLangRecordDestructure recordDeStmt) {
     @Override
     public void visit(BLangErrorDestructure errorDeStmt) {
         BLangErrorVarRef varRef = errorDeStmt.varRef;
-        if (varRef.message == null) {
-            dlog.error(varRef.pos, DiagnosticErrorCode.MISSING_REQUIRED_ARG_BINDING_PATTERN_ERROR_MESSAGE);
-        } else if (varRef.message.getKind() != NodeKind.SIMPLE_VARIABLE_REF ||
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2000]: lets remove this from `DiagnosticErrorCode.java` and `compiler.properties`

Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2000]: Fixed in 9526b8a

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -1638,15 +1638,17 @@ private void createVarDefStmts(BLangErrorVariable parentErrorVariable, BLangBloc
...
+            parentErrorVariable.message.expr = generateErrorMessageBuiltinFunction(parentErrorVariable.message.pos,
+                    parentErrorVariable.message.type, convertedErrorVarSymbol, null);
 
-        if (names.fromIdNode((parentErrorVariable.message).name) == Names.IGNORE) {
-            parentErrorVariable.message = null;
-        } else {
-            BLangSimpleVariableDef reasonVariableDef =
-                    ASTBuilderUtil.createVariableDefStmt(parentErrorVariable.message.pos, parentBlockStmt);
-            reasonVariableDef.var = parentErrorVariable.message;
+            if (names.fromIdNode((parentErrorVariable.message).name) == Names.IGNORE) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:1645]: I don't think we need this inner parenthesis here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -1638,15 +1638,17 @@ private void createVarDefStmts(BLangErrorVariable parentErrorVariable, BLangBloc
...
+                parentErrorVariable.message = null;
+            } else {
+                BLangSimpleVariableDef reasonVariableDef =
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:1648]: Shall we rename this to `messageVariableDef`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -2576,8 +2578,8 @@ private void createVarRefAssignmentStmts(BLangRecordVarRef parentRecordVarRef, B
 
     private void createVarRefAssignmentStmts(BLangErrorVarRef parentErrorVarRef, BLangBlockStmt parentBlockStmt,
                                              BVarSymbol errorVarySymbol, BLangIndexBasedAccess parentIndexAccessExpr) {
-        if (parentErrorVarRef.message.getKind() != NodeKind.SIMPLE_VARIABLE_REF ||
-                names.fromIdNode(((BLangSimpleVarRef) parentErrorVarRef.message).variableName) != Names.IGNORE) {
+        if (parentErrorVarRef.message != null && (parentErrorVarRef.message.getKind() != NodeKind.SIMPLE_VARIABLE_REF
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:2581]: I think we can remove `parentErrorVarRef.message.getKind() != NodeKind.SIMPLE_VARIABLE_REF` as it seems `message` is always a variable reference or null.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 6]
```diff
@@ -1995,14 +1999,14 @@ public void visit(BLangRecordDestructure recordDeStmt) {
...
-                names.fromIdNode(((BLangSimpleVarRef) varRef.message).variableName) != Names.IGNORE) {
-            setTypeOfVarRefInErrorBindingAssignment(varRef.message);
-        } else {
-            // set message var refs type to no type if the variable name is '_'
-            varRef.message.type = symTable.noType;
+        if (varRef.message != null) {
+            if (varRef.message.getKind() != NodeKind.SIMPLE_VARIABLE_REF ||
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2003]: Same as https://github.com/ballerina-platform/ballerina-lang/pull/28293/files#r569193341

File: tests/jballerina-unit-test/src/test/resources/test-src/error/error_test.bal [hunk 7]
```diff
@@ -371,6 +371,31 @@ function testUnionErrorTypeDescriptionInferring() {
...
+type SampleError error<SampleErrorData>;
+
+function testErrorBindingPattern() {
+    SampleError error(info=info,fatal=fatal) = error SampleError("Sample Error", info = "Detail Info",
+    fatal = true);
+    assertEquality(info, "Detail Info");
+    assertEquality(fatal, true);
+
+    error error(data=transactionData) = error("TransactionError", data={"A":"a", "B":"b"});
+    assertEquality(transactionData.toString(), "{\"A\":\"a\",\"B\":\"b\"}");
```
Comment (rdhananjaya) [inline tests/jballerina-unit-test/src/test/resources/test-src/error/error_test.bal:390]: Shall we move this test to `ErrorVariableDefinitionTest` class.

File: tests/jballerina-unit-test/src/test/resources/test-src/error/error_test.bal [hunk 8]
```diff
@@ -371,6 +371,31 @@ function testUnionErrorTypeDescriptionInferring() {
...
+
+    string i;
+    boolean b;
+    error(info=i,fatal=b) = error SampleError("Sample Error", info = "Some Info", fatal = false);
+    assertEquality(i, "Some Info");
+    assertEquality(b, false);
```
Comment (rdhananjaya) [inline tests/jballerina-unit-test/src/test/resources/test-src/error/error_test.bal:396]: As `ErrorVariableReferenceTest` is disabled at the moment, we will have to keep this part here.

**Review by rdhananjaya — APPROVED** [review]

**Review by rdulmina — APPROVED** [review]

**Pull request merged**